### PR TITLE
Added full CRUD for EMLI contacts

### PR DESCRIPTION
**Pull request by mmumby**
# Main
- Add in Core the ability to perform CRUD operations for EMLI users 
- MineSpace change from reading users (/src/constants/contacts.js) to fetch users from database.
- Include regional , major mines and general contacts.

# Other
- Creation of new table called EMLI_contact to substitute regional_contact table
- Seed data in new table

# How to test
In Core, a user can be added/edited under the Admin tab: "MineSpace EMLI Contacts"
Screen Shot 2021-12-01 at 11 47 43 AM

- Ensure generated documents are pulling the correct information from regional office:
Screen Shot 2021-12-01 at 11 53 05 AM
Screen Shot 2021-12-01 at 11 53 15 AM

In MineSpace, the contacts should be read as follow:
- Regional mine: Regional contacts based on the region mine are added and general contacts
image

- Major mine: Major mine contacts based on the region
image

# Notes
Screen Shot 2021-12-01 at 11 49 26 AM
Screen Shot 2021-12-01 at 11 49 44 AM
Screen Shot 2021-12-01 at 11 51 26 AM

Screen Shot 2021-12-01 at 11 49 03 AM

What contacts appear on each region: 
https://miro.com/app/board/uXjVOebUdk4=/
https://bcmines.atlassian.net/jira/software/c/projects/MDS/boards/1?modal=detail&selectedIssue=MDS-3885

### CODE REVIEW

File: migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+ALTER TABLE emli_contact OWNER TO mds;
+
+COMMENT ON TABLE emli_contact IS 'Contains EMLI contacts information.';
+
+--Create SEQUENCE
+DO
+$$
+BEGIN
+   EXECUTE format('
```
Comment (JustinMacaulay) [inline migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql]: I am not sure why this sequence stuff is needed, we haven't used it on any other tables before and as far as I know if you create a new serial column in PSQL it sets up all this behind the scenes...

File: migrations/sql/afterMigrate__1__seed_data.sql [hunk 2]
```diff
@@ -1571,38 +1571,76 @@ VALUES
...
+    deleted_ind
+  )
 VALUES
-    ('ROE', 'NE', 'MMD-PrinceGeorge@gov.bc.ca', '250 565-4240', '250 565-4328', '350-1011 4th Avenue', 'Prince George, B.C. V2L 3H9'),
-    ('ROE', 'NW', 'MMD-Smithers@gov.bc.ca', '250 847-7383', '250 847-7603', '2nd Floor, 3726 Alfred Avenue', 'Smithers, B.C. V0J 2N0'),
-    ('ROE', 'SC', 'MMD-Kamloops@gov.bc.ca', '250 371-3912', NULL, '2nd Floor, 441 Columbia Street', 'Kamloops, B.C. V2C 2T3'),
-    ('ROE', 'SE', 'MMD-Cranbrook@gov.bc.ca', '250 417-6134', NULL, '202-100 Cranbrook Street North', 'Cranbrook, B.C. V1C 3P9'),
-    ('ROE', 'SW', 'SouthwestMinesDivision@gov.bc.ca', '778 698-3649', '250 953-3878', 'PO Box 9395, Stn Prov Govt', 'Victoria, B.C. V8W 9M9')
-ON CONFLICT DO NOTHING;
+  (nextval('emli_contact_id_seq'),'MMO', null,null,null,'PermRecl@gov.bc.ca',null,null,null,null,true, false, false),
```
Comment (JustinMacaulay) [inline migrations/sql/afterMigrate__1__seed_data.sql]: As per the previous comment the Serial ID should be handling the nextval portion of this.

File: migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql [hunk 3]
```diff
@@ -0,0 +1,43 @@
+--Recreate regional_contact table as emli_contact with additional columns
+CREATE TABLE IF NOT EXISTS emli_contact
```
Comment (JustinMacaulay) [inline migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql]: Should this table also use the AuditMixin or do we care who is modifying them?

File: services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    mailing_address_line_1 = db.Column(db.String(254))
+    mailing_address_line_2 = db.Column(db.String(254))
+    first_name = db.Column(db.String(100))
+    last_name = db.Column(db.String(100))
+    is_major_mine = db.Column(db.Boolean, nullable=False, default=False)
+    is_general_contact = db.Column(db.Boolean, nullable=False, default=False)
+
+    emli_contact = db.relationship(
+        'EMLIContactType',
+        backref='emli_contact',
```
Comment (JustinMacaulay) [inline services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py:29]: I am not sure how valuable creating this back ref is. Just food for thought and not something that needs to be changed.

File: services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        return cls.query.filter_by(emli_contact_type_code=emli_contact_type_code).filter_by(
+            deleted_ind=False).first()
+
+    @classmethod
+    def find_EMLI_contact_by_id(cls, contact_id):
+        return cls.query.filter_by(contact_id=contact_id).filter_by(deleted_ind=False).first()
+
+    @classmethod
+    def find_EMLI_contacts_by_mine_region(cls, mine_region_code, is_major_mine=None):
+        general_contacts = cls.find_EMLI_general_contacts()
```
Comment (JustinMacaulay) [inline services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py:85]: I might need a walkthrough of this find_by method.

File: services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        mmo_contact = cls.find_major_mine_office()
+
+        if is_major_mine == True:
+            return cls.query.filter_by(
+                mine_region_code=mine_region_code, is_major_mine=is_major_mine).filter_by(
+                    deleted_ind=False).union(mmo_contact).union(general_contacts).all()
+        elif is_major_mine == False:
+            return cls.query.filter_by(
+                mine_region_code=mine_region_code, is_major_mine=is_major_mine).filter_by(
```
Comment (JustinMacaulay) [inline services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py:94]: in both of these the deleted_ind filter can be merged into the previous one to remove the additional use of filter_by

File: services/core-api/app/api/EMLI_contacts/resources/EMLI_contact.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        contact = EMLIContact.find_EMLI_contact_by_id(contact_id)
+        if not contact:
+            raise NotFound('Contact not found.')
+
+        contact.deleted_ind = True
+        current_app.logger.info(f'Deleting {contact}')
+
+        contact.save()
+
+        return contact
```
Comment (JustinMacaulay) [inline services/core-api/app/api/EMLI_contacts/resources/EMLI_contact.py]: Deletes should not return the deleted object. The return format should be:
return None, 204

File: services/core-api/app/api/EMLI_contacts/resources/EMLI_contact_list.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+from flask import request
+from app.api.utils.custom_reqparser import CustomReqparser
+from app.api.utils.access_decorators import requires_role_edit_emli_contacts, requires_any_of, VIEW_ALL, MINESPACE_PROPONENT
+from app.api.utils.resources_mixins import UserMixin
+from app.api.EMLI_contacts.models.EMLI_contact import EMLIContact
+from app.api.EMLI_contacts.models.EMLI_contact_type import EMLIContactType
+from app.api.EMLI_contacts.response_models import EMLI_CONTACT_MODEL
+
+
+class EMLIContactListResource(Resource, UserMixin):
```
Comment (JustinMacaulay) [inline services/core-api/app/api/EMLI_contacts/resources/EMLI_contact_list.py:13]: Again same comment as on the Project Summary Resource these two classes should and can be in the same resource file.

File: services/core-api/app/api/EMLI_contacts/resources/EMLI_contact_list.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        data = self.parser.parse_args()
+
+        contact_type = data.get('emli_contact_type_code', None)
+        contact_desc = EMLIContactType.find_contact_type(contact_type)
+
+        general_contacts = EMLIContact.find_EMLI_general_contacts()
+        mmo_contact = EMLIContact.find_EMLI_contact('MMO')
+        roe_contact = EMLIContact.find_EMLI_contact('ROE', data.mine_region_code)
+
+        unique_by_region = roe_contact.emli_contact_type_code
```
Comment (JustinMacaulay) [inline services/core-api/app/api/EMLI_contacts/resources/EMLI_contact_list.py]: Just for thought and I don't think it is an issue here but this type of validation can open this up to potential race conditions or simultaneous posts

File: services/core-api/app/api/mines/region/models/region.py [hunk 10]
```diff
@@ -14,12 +14,12 @@ class MineRegionCode(AuditMixin, Base):
     effective_date = db.Column(db.DateTime, nullable=False, server_default=FetchedValue())
     expiry_date = db.Column(db.DateTime)
 
-    regional_contacts = db.relationship('RegionalContact', lazy='select')
+    regional_contacts = db.relationship('EMLIContact', lazy='select')
```
Comment (JustinMacaulay) [inline services/core-api/app/api/mines/region/models/region.py:17]: Can you add a TODO above this at the least to rename it and all its uses for the future at the very least

File: services/core-api/app/api/now_submissions/resources/application_list_resource.py [hunk 11]
```diff
@@ -111,7 +111,7 @@ def _apply_filters_and_pagination(self,
     @api.doc(description='Save an application')
     @requires_role_edit_submissions
     @api.expect(APPLICATION)
-    @api.marshal_with(APPLICATION, code=201)
+    @api.marshal_with(APPLICATION, code=200)
```
Comment (JustinMacaulay) [inline services/core-api/app/api/now_submissions/resources/application_list_resource.py]: not sure why this code was changed, a 201 is a valid and expected code return for a post endpoint.

File: services/core-api/tests/status_code_gen.py [hunk 12]
```diff
@@ -39,6 +40,11 @@ def RandomMineRegionCode():
     return random.choice([x.mine_region_code for x in db.session.query(MineRegionCode).all()])
 
 
+def RandomEMLIContactTypeCode():
+    return random.choice(
+        [x.emli_contact_type_code for x in db.session.query(EMLIContactType).all()])
```
Comment (JustinMacaulay) [inline services/core-api/tests/status_code_gen.py]: Something we need to do on a lot of our factories but with this one can you filter it by active=True. this has caused issues in the past in some tests were codes have been marked as inactive and the code changed around that causing the test to fail.

File: services/core-web/common/actionCreators/minespaceActionCreator.js [hunk 13]
```diff
@@ -63,3 +63,86 @@ export const deleteMinespaceUser = (minespaceUserId) => (dispatch) => {
...
+    .then((response) => {
+      dispatch(success(reducerTypes.GET_EMLI_CONTACTS));
+      dispatch(minespaceActions.storeEMLIContacts(response.data));
+      return response;
+    })
+    .catch(() => dispatch(error(reducerTypes.GET_EMLI_CONTACTS)))
+    .finally(() => dispatch(hideLoading()));
+};
+
+export const fetchEMLIContactsByRegion = (region, isMajorMine) => (dispatch) => {
```
Comment (JustinMacaulay) [inline services/core-web/common/actionCreators/minespaceActionCreator.js:82]: is this only for region or major mine as well?

File: services/core-web/src/components/Forms/EMLIContacts/EMLIContactForm.js [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+            </Row>
+            <Row gutter={16}>
+              <Col span={24}>
+                <Form.Item>
+                  <Field
+                    id="mailing_address_line_2"
+                    name="mailing_address_line_2"
+                    label={
+                      props.formValues.emli_contact_type_code === regionalOfficeCode
+                        ? "Mailing Address Line 2*"
```
Comment (JustinMacaulay) [inline services/core-web/src/components/Forms/EMLIContacts/EMLIContactForm.js:262]: is there anywhere on this form where the asterisk used here is explained?

**Review by JustinMacaulay — COMMENTED** [review]
Looks good overall just a few comments.

File: migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql (hunk 3 above)
Comment (mmumby) [reply migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql]: oh yes we do care, good catch
@areyeslo lets add the audit mixins

File: services/core-web/src/components/Forms/EMLIContacts/EMLIContactForm.js (hunk 14 above)
Comment (mmumby) [reply services/core-web/src/components/Forms/EMLIContacts/EMLIContactForm.js:262]: No, The asterisk just means required. 
So the address is required for Regional offices, not for the MM office. Additionally, the address is only visible to the two office type codes. 

@JustinMacaulay does that answer your question?

File: services/core-web/common/actionCreators/minespaceActionCreator.js (hunk 13 above)
Comment (mmumby) [reply services/core-web/common/actionCreators/minespaceActionCreator.js:82]: MM as well - given the region and the MM flag, the API will return either the regional contacts based on region or the MM contacts based on region (and the standard MM contacts that apply to every region)

File: services/core-api/app/api/mines/region/models/region.py (hunk 10 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/region/models/region.py:17]: RegionalContact not longer exist and it was replaced in all its uses by EMLIContact. Should I add a note on this regard?

File: migrations/sql/afterMigrate__1__seed_data.sql (hunk 2 above)
Comment (areyeslo) [reply migrations/sql/afterMigrate__1__seed_data.sql]: Absolutely. This should be handled by the serial field.

File: migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql (hunk 1 above)
Comment (areyeslo) [reply migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql]: This is removed.

File: migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql (hunk 3 above)
Comment (mmumby) [reply migrations/sql/V2021.11.25.12.10__create_emli_contact_table.sql]: Added the audit columns!

File: services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py (hunk 5 above)
Comment (areyeslo) [reply services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py:85]: This function is used in minespace to populate the contact cards. If major mine, then we display the Major mine officer and major mine contact for the region. If regional mine, then we display the regional mine contacts and the general contacts who have are a major mine contact and do not have region (Chief Inspector of Mines, Chief Permitting Officer).

File: services/core-api/app/api/now_submissions/resources/application_list_resource.py (hunk 11 above)
Comment (areyeslo) [reply services/core-api/app/api/now_submissions/resources/application_list_resource.py]: Correct.

File: services/core-api/tests/status_code_gen.py (hunk 12 above)
Comment (areyeslo) [reply services/core-api/tests/status_code_gen.py]: Added, to ensure the record is an active one.

File: services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py (hunk 4 above)
Comment (areyeslo) [reply services/core-api/app/api/EMLI_contacts/models/EMLI_contact.py:29]: This is used to returned the contacts in a sorted way.

**Review by JustinMacaulay — COMMENTED** [review]
When tests are done and passing let me know and I will approve

**Review by JustinMacaulay — DISMISSED** [review]
Looks good and thanks for addressing everything.

**Review by JustinMacaulay — APPROVED** [review]

**Pull request merged**